cpp/src/arrow/ipc/reader.h

```diff
@@ -190,6 +193,85 @@ class ARROW_EXPORT RecordBatchFileReader {
   virtual Status ReadRecordBatch(int i, std::shared_ptr<RecordBatch>* batch) = 0;
 };
 
+/// \class RecordBatchReceiverCollect
+/// \brief Collect record batches read by RecordBatchStreamEmitter.
+class ARROW_EXPORT RecordBatchReceiverCollect : public Receiver {
+ public:
+  RecordBatchReceiverCollect() : record_batches_() {}
+  virtual ~RecordBatchReceiverCollect() = default;
+
+  Status RecordBatchReceived(std::shared_ptr<RecordBatch> record_batch) override {
+    record_batches_.push_back(std::move(record_batch));
+    return Status::OK();
+  }
+
+  /// \return the all read record batches
+  std::vector<std::shared_ptr<RecordBatch>> record_batches() const {
+    return record_batches_;
+  }
+
+ private:
+  std::vector<std::shared_ptr<RecordBatch>> record_batches_;
+};
+
+/// \class RecordBatchStreamEmitter
+/// \brief Push style record batch stream reader that receives data
+/// from user.
+///
+/// This class reads the schema (plus any dictionaries) as the first
+/// messages in the fed data, followed by record batches.
+class ARROW_EXPORT RecordBatchStreamEmitter {
+ public:
+  /// \brief Construct a record batch emitter.
+  ///
+  /// \param[in] receiver a Receiver that must implement
+  /// Receiver::MessageReceived to receive read record batches
+  /// \param[in] options any IPC reading options (optional)
+  RecordBatchStreamEmitter(Receiver* receiver,
```

Comment:
I think this definitely needs to be a `shared_ptr<Receiver>`. Async applications need robust lifetime handling.

Reply:
I'm unsure about this. Since both the Receiver as well as the emitter are internal classes, it seems likely that their lifetimes will be managed together. So if there _are_ lifetime considerations, then it seems more like that you would have a scenario like

```
class Something {

 private:
  std::shared_ptr<MyReceiver> my_receiver_;
  std::shared_ptr<RecordBatchStreamEmitter> emitter_;
};
```

I'm just thinking about the functions in this patch where an ephemeral emitter is created to parse a single IPC message, perhaps the extra overhead isn't meaningful

Reply:
I'm not sure why the Receiver would be an internal class. Any user of the C++ Arrow library may want to take advantage of event-based IPC stream decoding.